### PR TITLE
feat: add highlights like on godbolt

### PR DESCRIPTION
**Pull request by pinbraerts**
This PR implements source buffer highlighting which requires some additional state management

During the development I made some minor tweaks which could be divided to separate PRs if needed:
- bumped version of fennel
- `vim.tbl_extend(config, user_config)` instead of manual population
- highlight all line (>100 symbols)
- use `nvim_api` instead of `cmd` for augroups
- additional buffer options
- auto-cleanup buffers
- highlight group generator for hex colors

### CODE REVIEW

**Comment by pinbraerts** [comment]
Draft because I do not understand the stuff with offset. @p00f, can you please explain what does it mean?

**Comment by pinbraerts** [comment]
#21

**Comment by pinbraerts** [comment]
Reference
Reference

Current result
Current result

File: fnl/godbolt/assembly.fnl
Comment (dchinmay2) [inline fnl/godbolt/assembly.fnl]: use `&until` instead of a lua block: https://fennel-lang.org/reference#each-general-iteration

File: fnl/godbolt/assembly.fnl
Comment (dchinmay2) [inline fnl/godbolt/assembly.fnl]: did you mean the `if` to be inside the `let` block?

File: fnl/godbolt/assembly.fnl
Comment (dchinmay2) [inline fnl/godbolt/assembly.fnl]: can we not do this until absolutely needed? like if `nvim_create_autocmd` is the only neovim 0.7+ feature we're using, can we just stay with vimscript `autocmd` until we need more new features?

**Comment by dchinmay2** [comment]
i'll do a full review once you mark this as ready for review, feel free to ping when you do that

**Comment by dchinmay2** [comment]
> Draft because I do not understand the stuff with offset. @p00f, can you please explain what does it mean?

what are you referring to?

**Comment by dchinmay2** [comment]
re: "[Fennel 1.4.2 on PUC Lua 5.3](https://github.com/p00f/godbolt.nvim/pull/22/commits/a20c1c1a7f08dfaa54b5a4b07426d2006f7bf131)", please make sure it works on luajit

**Comment by dchinmay2** [comment]
Thanks for working on this!

**Comment by pinbraerts** [comment]
> what are you referring to?

Entries have an offset which was subtracted from the source line number. As I understand, when you pass a visual selection to godbolt, it outputs source line numbers inside the selection, so you have to use the offset. Am I right?

**Comment by pinbraerts** [comment]
> re: "[Fennel 1.4.2 on PUC Lua 5.3](https://github.com/p00f/godbolt.nvim/pull/22/commits/a20c1c1a7f08dfaa54b5a4b07426d2006f7bf131)", please make sure it works on luajit

Which version should I use and how to install it? I've installed mine via luarocks. Also could you share your fennel-ls configuration?

**Comment by pinbraerts** [comment]
> Entries have an offset which was subtracted from the source line number. As I understand, when you pass a visual selection to godbolt, it outputs source line numbers inside the selection, so you have to use the offset. Am I right?

I figured it out, tested by `:Godbolt`'ing a function and a whole file

offset

**Comment by pinbraerts** [comment]
@p00f, so it is ready. I'm open to splitting the PR if there is too much changes (mentioned in the first comment)

**Review by dchinmay2 — CHANGES_REQUESTED** [review]

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: Sure, I'll look at this

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: Sorry, I've just learned about fennel yesterday. I need to return the first matching value and I haven't managed to do this via `&until`

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: It is inside, just have to do formatting. Btw, how to do auto formatting is this repo?

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: > use `&until` instead of a lua block: https://fennel-lang.org/reference#each-general-iteration

The docs say that `each` returns `nil`. I may do this via `collect`, returning the first element, but it would be unreadable. I'm sure there has to be some canonical way to find something in a table.

Comment (dchinmay2) [reply fnl/godbolt/assembly.fnl]:
```fennel
(fn find-source [asm-buffer]
  (do
    (var source-buffer-ret nil)
    (each [source-buffer asm-buffers (pairs map) &until source-buffer-ret]
      (when (. asm-buffers asm-buffer)
        (set source-buffer-ret source-buffer)))
    source-buffer-ret))

```

you also don't have to compare `(. asm-buffers asm-buffer)` with `nil`, `nil` is falsey

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: Done, everything works. But I could used more functions from newer nvim api elsewhere

File: fnl/godbolt/assembly.fnl
Comment (dchinmay2) [reply fnl/godbolt/assembly.fnl]: there's a script: https://github.com/p00f/godbolt.nvim/blob/main/format

you need fnlfmt: https://git.sr.ht/~technomancy/fnlfmt

**Comment by dchinmay2** [comment]
Sorry, I might need a day or two to look at this properly

File: fnl/godbolt/assembly.fnl
Comment (dchinmay2) [reply fnl/godbolt/assembly.fnl]: right sorry, you just use `make format`, there's a makefile which calls this script with the filenames

File: fnl/godbolt/assembly.fnl
Comment (dchinmay2) [reply fnl/godbolt/assembly.fnl]: Where?

Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: Just built nvim 0.6.0 to check. [`nvim_create_user_command`](https://github.com/neovim/neovim/blob/5581a95534e44b8714e715c925c9de2d95ae1c21/src/nvim/api/command.c#L880) is used [here](https://github.com/p00f/godbolt.nvim/blob/09b78f294ea26af678a9ac363b7d5ce8e20b696e/fnl/godbolt/macros.fnl#L18). This method was introduced in nvim 0.9

Comment (dchinmay2) [reply fnl/godbolt/assembly.fnl]: ah well, since nobody complained yet, we can use 0.9 apis

Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: So should I revert changes to calling lua functions?

I was thinking about implementing backports to nvim apis in a separate plugin

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: This is old version of the function. Formatted the code

File: fnl/godbolt/assembly.fnl
Comment (pinbraerts) [reply fnl/godbolt/assembly.fnl]: Done in https://github.com/p00f/godbolt.nvim/pull/22/commits/0d3b09db70033cd30a17da85431cb5f041d06efd and tested locally

**Comment by dchinmay2** [comment]
Sorry, I'll look at this soon

**Comment by dchinmay2** [comment]
I get this:
```
Error detected while processing CursorMoved Autocommands for "<buffer=1>":
E5248: Invalid character in group name
```

**Comment by pinbraerts** [comment]
Are you using the latest commit? Try doing `make clean && make format && make`. Which version of neovim are you using?

**Comment by dchinmay2** [comment]
`aa6b9c677d`

**Comment by pinbraerts** [comment]
https://github.com/p00f/godbolt.nvim/commit/2e42a19ed56b3355de8f879243dae95584c42b91
Is the latest

**Comment by dchinmay2** [comment]
I meant neovim commit - I am using that commit of your pr

**Comment by pinbraerts** [comment]
> I meant neovim commit - I am using that commit of your pr

Ok, I'll test on the older version. I am using v0.10.0

**Comment by pinbraerts** [comment]
nvim v0.9.0 works fine

Built nvim with your commit, cant reproduce the error either

**Comment by pinbraerts** [comment]
Ok, I reproduced the error on a vm with your nvim version, debugging

**Comment by pinbraerts** [comment]
> I get this:
> ```
> Error detected while processing CursorMoved Autocommands for "<buffer=1>":
> E5248: Invalid character in group name
> ```

You have to setup the plugin with `config = true` or `opts = {}` in lazy, the problem goes away

Minimal `init.lua`:
```lua
-- Bootstrap lazy.nvim
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not (vim.uv or vim.loop).fs_stat(lazypath) then
  local lazyrepo = "https://github.com/folke/lazy.nvim.git"
  vim.fn.system({ "git", "clone", "--filter=blob:none", "--branch=stable", lazyrepo, lazypath })
end
vim.opt.rtp:prepend(lazypath)

require("lazy").setup({
	{
		"pinbraerts/godbolt.nvim",
		branch = "highlights",
		config = true,
		-- opts = {},
	},
})
```

**Comment by dchinmay2** [comment]
Config must remain optional, please find another way.

**Comment by pinbraerts** [comment]
> Config must remain optional, please find another way.

Well, I initialize a highlight groups in the setup. Should I do this lazily or completely remove the code and shift the task to the user side like [here](https://github.com/HiPhish/rainbow-delimiters.nvim)? Still, I want to have a default value.

I suggest the following spec. The highlighting is a config parameter which could be set to:
1. `false` to completely disable
2. `true` to enable default method
3. `string` to enable only visual and set group for it
4. `array[string]` to set custom highlight groups, first in the array is used for visual

Optional, could be done on user side:
5. `array[string]` if an entry starts with `'#'`, automatically define groups with the color

**Comment by dchinmay2** [comment]
> Well, I initialize a highlight groups in the setup.

Do it in `plugin/godbolt.fnl`

**Comment by dchinmay2** [comment]
See `:h add-global-plugin`

**Comment by dchinmay2** [comment]
In `plugin/godbolt.fnl`, add calls to `vim.api.nvim_set_hl` to define the highlight groups

~~Or you could use vimscript `plugin/godbolt.vim` if that feels more natural for defining highlight groups~~ scratch this, we already have `plugin/godbolt.fnl`

**Comment by dchinmay2** [comment]
Again, this does not need to be done via a configuration table, the user/colorschemes can simply override the highlight groups in their dotfiles if they want to configure it

I apologise if it feels I'm being nitpicky, this PR is great work, I just want to push back against the mandatory `.setup()` paradigm

**Comment by dchinmay2** [comment]
Alright, should I review now?

**Comment by pinbraerts** [comment]
> Alright, should I review now?

Yes

**Comment by pinbraerts** [comment]
> I apologise if it feels I'm being nitpicky, this PR is great work, I just want to push back against the mandatory `.setup()` paradigm

Totally valid suggestions, no need to apologise

**Comment by pinbraerts** [comment]
The PR is rather experimental and I'm open to discussing the specification and improving flexibility here or in the next PRs.

If you want to know my opinion: I'm against polluting the global namespace because as a user I want to have full control over the plugin. Ideally it should not execute any code I didn't ask it to via `setup`. Also, I am very pro-lazy, so almost every plugin I use is not loaded unless required (like mason and tree-sitter). This results in instant launch and no errors on startup. As soon as lazy doesn't execute `godbolt.fnl` on startup if the plugin is configured to load only on `:Godbolt`, I'm ok with this.

**Comment by dchinmay2** [comment]
My opinion: this obsession with complete laziness is not productive, the plugin already lazy loads itself, only defining the commands. The actual code is only loaded once you `:Godbolt`. This is how traditional vimscript plugins worked (and still work) - the `plugin` folder defines commands and highlight groups, and the `autoload` folder defines functions which are sourced on-demand. The lua equivalent of the `autoload` folder is `lua`.

This approach removes the need to add anything to the user's configuration unless they want to load, while maintaining speed.

Side note: this is why I dislike lazy.nvim, it promotes the `.setup()` culture

Re: global namespace, I will fix that in a refactor

**Comment by dchinmay2** [comment]
Do you mind me pushing to your branch?

**Comment by pinbraerts** [comment]
> Do you mind me pushing to your branch?

Don't mind

**Comment by pinbraerts** [comment]
> My opinion: this obsession with complete laziness is not productive, the plugin already lazy loads itself, only defining the commands.

Yes, it is sometimes unproductive, especially if every plugin does the same, but differently. But isn't it more unproductive to reinvent the same bike of lazy loading again and again in every plugin, rather than adding a line in README with suggestions on lazy loading?

> This approach removes the need to add anything to the user's configuration unless they want to load, while maintaining speed.

I totally understand you, but whether your plugin is fast or not is defined by the user. Even if every plugin is only defining it's cmd's lazily, it still adds up to startup time and could become unbearable

> Side note: this is why I dislike lazy.nvim, it promotes the `.setup()` culture

I would seek to do the same in packer or any other plugin manager

To sum up, if I ever make a plugin, I would assume lazy.nvim/packer community much better than me in implementing lazy-loading in the plugin, even if it looks relatively simple. The same thing about old nvim versions and newer apis. I would like a single solution which is used by many plugins rather than implementing backports myself in every plugin. I might start a project, if I stumble upon the same thing again)

No offense, it's just an opinions. I really appreciate your opinion and admire your contribution to the community

**Comment by dchinmay2** [comment]
> I would assume lazy.nvim/packer community much better than me in implementing lazy-loading in the plugin

The plugin author knows about their plugin more than a plugin manager 😛

I've pushed some fixes to `pinbraerts-highlights` on this repo. I'll merge tomorrow after seeing if `update-hl` can be simplified, adding docs, etc

**Comment by pinbraerts** [comment]
> I've pushed some fixes to `pinbraerts-highlights` on this repo. I'll merge tomorrow after seeing if `update-hl` can be simplified, adding docs, etc
 
Good, let me know if something is unclear

**Comment by pinbraerts** [comment]
The edits in my branch are allowed

Screenshot_20240630-093410

**Comment by dchinmay2** [comment]
I'm trying to see if we can avoid clearing and re-applying the static multi-colour highlights when the cursor moves

**Comment by pinbraerts** [comment]
> I'm trying to see if we can avoid clearing and re-applying the static multi-colour highlights when the cursor moves

Added suggestions for that in [`pinbraerts/hi`](https://github.com/pinbraerts/godbolt.nvim/tree/hi):[`222ff88`](https://github.com/pinbraerts/godbolt.nvim/commit/222ff8841128f053116b5efa2cb240300709bdfa)

**Comment by dchinmay2** [comment]
Haven't actually tried this, but if you move the cursor, won't the previous line be unhighlighted? (instead of being highlighted with the static colour)

**Comment by pinbraerts** [comment]
> Haven't actually tried this, but if you move the cursor, won't the previous line be unhighlighted? (instead of being highlighted with the static colour)

I thought so as well. But it works, I've put them in different namespaces, maybe that's the thing. I also logged inside ifs to see if my code is actually executing

**Comment by dchinmay2** [comment]
I see, I'll do this within 1 day I promise

**Comment by pinbraerts** [comment]
> I see, I'll do this within 1 day I promise

I will push all the changes here before merging to track progress, if you don't mind

**Comment by dchinmay2** [comment]
Sure

**Comment by dchinmay2** [comment]
I was thinking maybe we could have a separate function for the static multicolour highlights which would be called only once, it would make things simpler

**Comment by pinbraerts** [comment]
> Haven't actually tried this, but if you move the cursor, won't the previous line be unhighlighted? (instead of being highlighted with the static colour)

Previously the line was unhighlighted because of multiple highlights on the same line. I added a list of highlighted lines so that one line is never highlighted twice.

multiple highligts

correct

**Comment by pinbraerts** [comment]
> I was thinking maybe we could have a separate function for the static multicolour highlights which would be called only once, it would make things simpler

Done

**Comment by dchinmay2** [comment]
Looks good, can you add docs?

**Comment by pinbraerts** [comment]
> Looks good, can you add docs?

Added docs, refactored config parameters a bit

**Comment by dchinmay2** [comment]
Should I merge now?

**Comment by pinbraerts** [comment]
> Should I merge now?

Yes, you can review, test and merge

**Comment by dchinmay2** [comment]
I renamed the field from `lines` to static, `static = false` doesn't seem to work

**Comment by pinbraerts** [comment]
> I renamed the field from `lines` to static, `static = false` doesn't seem to work

Revert the changes in `setup()`. Assigning to `config` apparently just overwrites the local reference, not the table itself

**Comment by pinbraerts** [comment]
Also you've committed a cpp file

**Comment by pinbraerts** [comment]
Cursor can be specified by hex color too

**Comment by dchinmay2** [comment]
But I do the same in https://github.com/p00f/clangd_extensions.nvim/blob/a8500531c4ed3a207e744a374ea038744a0f93eb/lua/clangd_extensions/config.lua#L53 and it works

I might have to look at how lua works again

**Comment by pinbraerts** [comment]
> But I do the same in https://github.com/p00f/clangd_extensions.nvim/blob/a8500531c4ed3a207e744a374ea038744a0f93eb/lua/clangd_extensions/config.lua#L53 and it works
> 
> I might have to look at how lua works again

That's totally different. You set a field of an object in a function. But here you return an object and reset local reference, not the field in an object. But I'm not totally sure. You can do the same with introducing `M` here

**Comment by dchinmay2** [comment]
Done, thanks for your contribution and patience!

**Pull request merged**
